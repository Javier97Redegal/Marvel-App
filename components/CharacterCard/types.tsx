type CharacterCardProps = {
    character: {
        id: number
        name: string | undefined
        thumbnail: {
            path: string
            extension: string
        }
    }
}