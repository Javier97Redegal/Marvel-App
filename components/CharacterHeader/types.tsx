export interface CharacterHeaderProps {
    id: number
}