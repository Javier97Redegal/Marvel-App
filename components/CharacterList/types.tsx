export interface CharacterListProps {
    mode: string
    query: string
}