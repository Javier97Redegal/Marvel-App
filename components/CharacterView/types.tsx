export interface CharacterViewProps {
    id: string
}