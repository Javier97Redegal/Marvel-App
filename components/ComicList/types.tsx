export interface ComicListProps {
    id: number
}