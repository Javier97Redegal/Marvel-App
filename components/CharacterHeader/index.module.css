.header {
    position: relative;
    background: black;
    color: white;
}

.header::after {
    content: '';
    display: block;
    position: absolute;
    right: 0;
    bottom: 0;
    width: 0;
    height: 0;
    border-bottom: 40px solid white;
	border-left: 40px solid transparent;
}

.header .inner {
    padding: 3rem;
    max-width: 1980px;
    margin: 0 auto;
    display: flex;
    align-items: center;
    justify-content: space-between;
    gap: 3rem;

    @media (max-width: 800px) {
        flex-flow: column;
        gap: 2rem;
        padding: 1rem;
    }
}

.image {
    width: 40%;
    aspect-ratio: 1 / 1;
    background-size: cover;
    background-position: center;
    margin: -3rem 0;

    @media (max-width: 800px) {
        width: calc(100% + 2rem);
        margin: -1rem -1rem 0 -1rem;
    }
}

.body {
    width: 100%;
    font-size: 1.5rem;
    line-height: 2.5rem;

    @media (max-width: 1280px) {
        font-size: 1rem;
        line-height: 2rem;
    }
}

.body h1 {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin: 0;
    text-transform: uppercase;
    letter-spacing: 1px;
}

.body h1 button {
    display: block;
    width: 40px;
    background: transparent;
    border: none;
    border-radius: 0;
    cursor: pointer;
    flex: 0 0 auto;

    @media (max-width: 1280px) {
        width: 35px;
    }
}

.body h1 button img {
    display: block;
    width: 100%;
}