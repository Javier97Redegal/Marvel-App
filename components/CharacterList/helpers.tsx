export const getResultsMessage = (number: number): string => {
    return `Result${number === 1 ? '' : 's'}`
}