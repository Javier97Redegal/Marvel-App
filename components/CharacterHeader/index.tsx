import { useQuery } from 'react-query'
import { getCharacter } from '@/utils/MarvelAPI'
import { handleFavoriteButton } from '@/helpers'
import { useApp } from '@/context/AppContext'
import LoadingPanel from '@/components/LoadingPanel'
import styles from './index.module.css'
import { CharacterHeaderProps } from './types'

const CharacterHeader = ({
    id
}: CharacterHeaderProps) => {
    const { favorites, addFavorite, removeFavorite } = useApp()
    const fetchCharacter = async () => await getCharacter({ id })
    const { data, status } = useQuery(`character-${id}`, fetchCharacter)
    const isFavorite = favorites.map(favorite => favorite.id).includes(id)

    if (status !== 'success') return <LoadingPanel type={status} />

    const character = data.data.results[0]

    return <header className={styles.header}>
        <div className={styles.inner}>
            <div className={styles.image} style={{ backgroundImage: `url(${character.thumbnail.path}.${character.thumbnail.extension})` }}></div>
            <div className={styles.body}>
                <h1>
                    {character.name}
                    <button onClick={() => handleFavoriteButton({ character, isFavorite, addFavorite, removeFavorite })}>
                        {isFavorite ? <img src={'/heart-default.png'} alt='Remove from Favorites' title='Remove from Favorites' /> : <img src={'/heart-alt.png'} alt='Add to Favorites' title='Add to Favorites' />}
                    </button>
                </h1>
                <p>{character.description}</p>
            </div>
        </div>
    </header>
}

export default CharacterHeader