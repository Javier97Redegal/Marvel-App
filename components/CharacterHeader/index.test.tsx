import '@testing-library/jest-dom'
import { fireEvent, render, screen } from '@testing-library/react'
import { useQuery } from 'react-query'
import CharacterHeader from '.'
import { CharacterHeaderProps } from '@/components/CharacterHeader/types'
import { useApp } from '@/context/AppContext'
import { handleFavoriteButton } from '@/helpers'

jest.mock('react-query', () => ({
    useQuery: jest.fn()
}))

jest.mock('@/utils/MarvelAPI', () => ({
    getCharacter: jest.fn()
}))

jest.mock('@/helpers/index.tsx', () => ({
    handleFavoriteButton: jest.fn()
}))

jest.mock('@/context/AppContext', () => ({
    useApp: jest.fn()
}))

jest.mock('@/components/LoadingPanel', () => jest.fn(({ type }) => <div>LoadingPanel Mock: {type}</div>))

describe('CharacterHeader tests', () => {
    const mockProps: CharacterHeaderProps = { id: 12345 }
    const characterData = {
        data: {
            results: [
                {
                    id: '12345',
                    name: 'Spider-Man',
                    description: 'Friendly neighborhood Spider-Man.',
                    thumbnail: {
                        path: 'path/to/spiderman',
                        extension: 'jpg'
                    }
                }
            ]
        }
    }

    const mockUseApp = {
        favorites: [{ id: '12345' }],
        addFavorite: jest.fn(),
        removeFavorite: jest.fn()
    }

    beforeEach(() => {
        (useApp as jest.Mock).mockReturnValue(mockUseApp)
    })

    it('renders loading panel when loading', () => {
        (useQuery as jest.Mock).mockReturnValue({ data: null, status: 'loading' })

        const { asFragment } = render(<CharacterHeader {...mockProps} />)

        expect(screen.getByText('LoadingPanel Mock: loading')).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    it('renders loading panel when error', () => {
        (useQuery as jest.Mock).mockReturnValue({ data: null, status: 'error' })

        const { asFragment } = render(<CharacterHeader {...mockProps} />)

        expect(screen.getByText('LoadingPanel Mock: error')).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    it('renders character details when data is loaded', () => {
        (useQuery as jest.Mock).mockReturnValue({ data: characterData, status: 'success' })

        const { asFragment } = render(<CharacterHeader {...mockProps} />)

        expect(screen.getByText('Spider-Man')).toBeInTheDocument()
        expect(screen.getByText('Friendly neighborhood Spider-Man.')).toBeInTheDocument()
        expect(screen.getByRole('button')).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    it('handles favorite button click', () => {
        (useQuery as jest.Mock).mockReturnValue({ data: characterData, status: 'success' })

        render(<CharacterHeader {...mockProps} />)

        const button = screen.getByRole('button')
        fireEvent.click(button)

        expect(handleFavoriteButton).toHaveBeenCalled()
    })
})