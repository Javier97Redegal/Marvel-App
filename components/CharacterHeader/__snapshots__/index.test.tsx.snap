// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterHeader tests renders character details when data is loaded 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="inner"
    >
      <div
        class="image"
        style="background-image: url(path/to/spiderman.jpg);"
      />
      <div
        class="body"
      >
        <h1>
          Spider-Man
          <button>
            <img
              alt="Add to Favorites"
              src="/heart-alt.png"
              title="Add to Favorites"
            />
          </button>
        </h1>
        <p>
          Friendly neighborhood Spider-Man.
        </p>
      </div>
    </div>
  </header>
</DocumentFragment>
`;

exports[`CharacterHeader tests renders loading panel when error 1`] = `
<DocumentFragment>
  <div>
    LoadingPanel Mock: error
  </div>
</DocumentFragment>
`;

exports[`CharacterHeader tests renders loading panel when loading 1`] = `
<DocumentFragment>
  <div>
    LoadingPanel Mock: loading
  </div>
</DocumentFragment>
`;
